src/net_processing.cpp

```diff
@@ -381,6 +382,15 @@ struct Peer {
     /** Time of the last getheaders message to this peer */
     NodeClock::time_point m_last_getheaders_timestamp{};
 
+    /** Protects m_headers_sync **/
+    Mutex m_headers_sync_mutex;
+    /** Headers-sync state for this peer (eg for initial sync, or syncing large
+     * reorgs) **/
```

Comment:
IIUC, syncing large reorgs is defined here by if we receive chain headers forking more than the 144 block buffer as documented in `GetAntiDoSWorkThreshold()`. Any fork under that threshold should have its headers accepted without the anti-DoS headers mechanism playing out.